### ISSUE TITLE
[Issue] `pyls` detects venv too early

### ISSUE THREAD

**Issue by dbischof90** [neovim/nvim-lspconfig#500]
### Extended infos

- `nvim --version`: NVIM v0.5.0-dev+975-ga58c5509d
- nvim-lsp version(commit hash): 23a944c
- `:checkhealth` result:
```

health#lspconfig#check
========================================================================
## Checking language server protocol configuration
  - INFO: pyls: configuration checked.
  - INFO: rust_analyzer: configuration checked.
  - INFO: texlab: configuration checked.

health#completion_nvim#check
========================================================================
## general
  - OK: neovim version is supported

## completion source
  - OK: all completion source are valid

## snippet source
  - INFO: You haven't setup any snippet source.

health#vimtex#check
========================================================================
## vimtex
  - OK: Vim version should have full support!
  - OK: General viewer should work properly!
  - WARNING: Compiler callbacks will not work!
    - ADVICE:
      - `neovim-remote` / `nvr` is required for callbacks to work with neovim
      - Please also set |g:vimtex_compiler_progname| = 'nvr'
      - For more info, see :help |vimtex-faq-neovim|

health#nvim#check
========================================================================
## Configuration
  - OK: no issues found

## Performance
  - OK: Build type: RelWithDebInfo

## Remote Plugins
  - OK: Up to date

## terminal
  - INFO: key_backspace (kbs) terminfo entry: key_backspace=^H
  - INFO: key_dc (kdch1) terminfo entry: key_dc=\E[3~
  - INFO: $VTE_VERSION='5602'
  - INFO: $COLORTERM='truecolor'

health#provider#check
========================================================================
## Clipboard (optional)
  - OK: Clipboard tool found: xclip

## Python 2 provider (optional)
  - WARNING: No Python executable found that can `import neovim`. Using the first available executable for diagnostics.
  - ERROR: Python provider error:
    - ADVICE:
      - provider/pythonx: Could not load Python 2:
          python2 not found in search path or not executable.
          python2.7 not found in search path or not executable.
          python2.6 not found in search path or not executable.
          /home/daniel/.cache/pypoetry/virtualenvs/craes-pyclient-KENhYm7S-py3.9/bin/python is Python 3.9 and cannot provide Python 2.
  - INFO: Executable: Not found

## Python 3 provider (optional)
  - INFO: Using: g:python3_host_prog = "/usr/bin/python"
  - INFO: Executable: /usr/bin/python
  - INFO: Python version: 3.9.1
  - INFO: pynvim version: 0.4.2
  - OK: Latest pynvim is installed.

## Python virtualenv
  - INFO: $VIRTUAL_ENV is set to: /home/daniel/.cache/pypoetry/virtualenvs/craes-pyclient-KENhYm7S-py3.9
  - INFO: Python version: 3.9.1
  - OK: $VIRTUAL_ENV provides :!python.

## Ruby provider (optional)
  - WARNING: `ruby` and `gem` must be in $PATH.
    - ADVICE:
      - Install Ruby and verify that `ruby` and `gem` commands work.

## Node.js provider (optional)
  - INFO: Node.js: v15.5.0
  - WARNING: Missing "neovim" npm (or yarn) package.
    - ADVICE:
      - Run in shell: npm install -g neovim
      - Run in shell (if you use yarn): yarn global add neovim

## Perl provider (optional)
  - ERROR: perl provider error:
    - ADVICE:
      - "Neovim::Ext" cpan module is not installed
```
- What language server(If the problem is related to a specific language server): `pyls`
- Operating system/version: Arch Linux 5.9.14-arch1-1
- Terminal name/version: termite
- `$TERM`: `xterm-256color`

### How to reproduce the problem from neovim startup
Create an empty `poetry` project, add one arbitrary package, start nvim with a python file and import the package.

### Actual behaviour
`pyls` does not recognize the package and shows error lints.

### Expected behaviour
`pyls` should use the venv.

---------------------------------

I'm using the package `'petobens/poet-v'` to automatically manage my venv. Before I used `'w0rp/ale'` to manage my lints and LSPs. However, `pyls` seems to be loaded from my 'main' environment here.

As seen above, neovim seems to correctly detect my venv and in my `init.vim` I load `'petobens/poet-v'` at the very beginning. This was the trick for me to get the `ale` LSP to source the lsp from my venv instead of my main installation.

It seems like not with the lua config, this mechanic is broken. What can I do to resolve this?

**Comment by mjlbach** [comment]
Is pyls installed in your virtual environment? In my experience pyls only provides completions when installed into the virtualenv that contains said package. Can you reproduce with pyright?

**Comment by dbischof90** [comment]
Yes, I can reproduce the same:

1) pyls is installed in both the venv and in main. If I manually activate the shell before I launch neovim, the 'correct' pyls is called.
2) I see that the same happens with pyright.

Is there a way to debug to check which python venv is actuve when the lua setup function is called?

**Comment by mjlbach** [comment]
I believe the way it's working (or failing to work) is that neovim will launch the first pyls on your path, and probably whatever mechanism poet-v is using doesn't/can't override the search path that we're using when calling cmd. I haven't used poet-v, but I would recommend looking through their implementation to see what exactly they are updating. I can take a look at this, but it's going to be on the backburner for awhile.

**Comment by dbischof90** [comment]
That would be nice, I'm not sure how vim packages are loaded, `poet-v` seems to detect the necessary `$VIRTUAL_ENV` variable and sets that. However, I'm not sure when neovim executes this and whether that is loaded before or after executing the lua config. I know about `nvim -V` but I think I can't see the right order in there.

**Comment by mjlbach** [comment]
I think the primary issue is that poet-v should override the cmd we pass for the default_config to launch the language server. I don't know how willing they are to add upstream functionality specific to different language client implementations though

**Comment by dbischof90** [comment]
I see that in the `pyls` lsp one can pass an executable. That could be wrapped with a bash script. Is that possible for the other LSPs too?

**Comment by mjlbach** [comment]
I think the easier solution is to just set the local variables PYTHON_PATH, PATH, etc in the cmd

**Comment by mjlbach** [comment]
Ok. Here is the issue.

1. Make two virtualenvs:
```bash
python3 -m venv $HOME/.virtualenvs/test1
python3 -m venv $HOME/.virtualenvs/test2
source $HOME/.virtualenvs/test1/bin/activate
pip install pillow
source $HOME/.virtualenvs/test2/bin/activate
pip install requests
```
2. Make a file `$HOME/test.sh` with the following contents
```bash
VIRTUAL_ENV=$HOME/.virtualenvs/test1 $HOME/.virtualenvs/test1/bin/pyls
```
3. Make it executable `chmod +x $HOME/test.sh`
4. Configure pyls to use the test script
```lua
nvim_lsp.pyls.setup {
  cmd = {"$HOME/test.sh"};
  on_attach = on_attach;
}
```
5. Try changing between 1 and 2 for both VIRTUAL_ENV and test#/bin/pyls while doing completions on the following file
```python
import requests
import PIL

def main() {
    requests.get()
    PIL.Image()
}
```

The only things that matters is the VIRTUAL_ENV prefix to launching pyls. You can't set this as cmd since we check for executables as the first argument. I would recommend you make a PR injecting environmental variables into the launch environment of cmd. Then the author of poet-v can add support if they choose.

**Comment by mjlbach** [comment]
This should already be supported 
https://github.com/mjlbach/neovim/blob/eaee3d9297d114c81b05dd4ed104373eeb87bc11/runtime/lua/vim/lsp.lua#L334-L340

**Comment by keitwb** [comment]
@mjlbach Is there any way to dynamically set that `cmd_env` in your example above based on the root dir resolved for the current file?  I have a monorepo with multiple python apps, each with their own virtualenv dir and often flip between them in the same nvim session.

**Comment by lithammer** [comment]
@keitwb Can't speak for pyls, but for pyright you can do something like this:

```lua
local configs = require('lspconfig/configs')
local util = require('lspconfig/util')

local path = util.path

local function get_python_path(workspace)
  -- Use activated virtualenv.
  if vim.env.VIRTUAL_ENV then
    return path.join(vim.env.VIRTUAL_ENV, 'bin', 'python')
  end

  -- Find and use virtualenv in workspace directory.
  for _, pattern in ipairs({'*', '.*'}) do
    local match = vim.fn.glob(path.join(workspace, pattern, 'pyvenv.cfg'))
    if match ~= '' then
      return path.join(path.dirname(match), 'bin', 'python')
    end
  end

  -- Fallback to system Python.
  return exepath('python3') or exepath('python') or 'python'
end

lspconfig.pyright.setup({
  -- ...
  before_init = function(_, config)
    config.settings.python.pythonPath = get_python_path(config.root_dir)
  end
})
```

Then you can expand the `get_python_path` with support for Poetry or Pipenv if that's you cup of tea. Or whatever style your monorepo has.

**Comment by gjeusel** [comment]
Here is my version for pylsp, inspired by the one from lithammer, using the  lspconfig `on_new_config` hook, for conda + poetry setup:

```lua
local lspconfig = require("lspconfig")
local lspconfig_util = require('lspconfig/util')
local telescope_utils = require('telescope.utils')

local path = lspconfig_util.path
local get_os_command_output = telescope_utils.get_os_command_output

local function file_exists(name)
  local f = io.open(name, 'r')
  if f ~= nil then
    io.close(f)
    return true
  else
    return false
  end
end

local function get_python_path(workspace)
  -- https://github.com/neovim/nvim-lspconfig/issues/500#issuecomment-851247107

  -- If is poetry project, find the associated venv and use it
  -- TODO: find a faster way to do so, as it takes:
  --       ~ 0.3 seconds for conda run -n base
  --       ~ 0.3 seconds for poetry run which python
  if file_exists(path.join(workspace, 'poetry.lock')) then
    -- Make sure to not mess up with potential activated conda env
    local poetry_cmd = {'poetry', 'run', 'which', 'python'}
    if vim.env.CONDA_PREFIX then
      local cmd = {'conda', 'run', '-n', 'base', unpack(poetry_cmd)}
      local stdout, ret = get_os_command_output(cmd, workspace)
      if ret == 0 and #stdout > 1 then
        return stdout[#stdout - 1]
      end
    else
      local stdout, ret = get_os_command_output(poetry_cmd, workspace)
      if ret == 0 and #stdout ~= 0 then
        return stdout[#stdout]
      end
    end
  end

  -- If conda env is activated, use it
  if vim.env.CONDA_PREFIX then
    return path.join(vim.env.CONDA_PREFIX, 'bin', 'python')
  end

  -- If virtualenv is activated, use it
  if vim.env.VIRTUAL_ENV then
    return path.join(vim.env.VIRTUAL_ENV, 'bin', 'python')
  end

  -- Fallback to system Python.
  return 'python'
end

lspconfig.pylsp.setup({
  -- ...
  on_new_config = function(new_config, new_root_dir)
    local python_path = get_python_path(new_root_dir)
    new_config.cmd = {python_path, '-m', 'pylsp'}
  end
})
```

Warning: It's a bit slow on startup (add 0.6 seconds in the worst case)

**Comment by sagarc03** [comment]
> @keitwb Can't speak for pyls, but for pyright you can hook into `before_init` to do something like this:
> 
> ```lua
> local configs = require('lspconfig/configs')
> local util = require('lspconfig/util')
> 
> local path = util.path
> 
> local function get_python_path(workspace)
>   -- Use activated virtualenv.
>   if vim.env.VIRTUAL_ENV then
>     return path.join(vim.env.VIRTUAL_ENV, 'bin', 'python')
>   end
> 
>   -- Find and use virtualenv in workspace directory.
>   for _, pattern in ipairs({'*', '.*'}) do
>     local match = vim.fn.glob(path.join(workspace, pattern, 'pyvenv.cfg'))
>     if match ~= '' then
>       return path.join(path.dirname(match), 'bin', 'python')
>     end
>   end
> 
>   -- Fallback to system Python.
>   return exepath('python3') or exepath('python') or 'python'
> end
> 
> lspconfig.pyright.setup({
>   -- ...
>   before_init = function(_, config)
>     config.settings.python.pythonPath = get_python_path(config.root_dir)
>   end
> })
> ```
> 
> Then you can expand the `get_python_path` function with support for Poetry or Pipenv if that's you cup of tea. Or whatever style your monorepo has.

Thanks for the insight, but I have tried this multiple times the `before_init` function is not being called. Tried changing it to `on_init` it seems to work that way but pyright still can't resolve most of the imports.

**Comment by gzbd** [comment]
I tweaked @lithammer code to the below one. My use case is that I want to use venv created by Pipenv without explicitly jumping to it before starting neovim.

```
local util = require('lspconfig/util')

local path = util.path

local function get_python_path(workspace)
  -- Use activated virtualenv.
  if vim.env.VIRTUAL_ENV then
    return path.join(vim.env.VIRTUAL_ENV, 'bin', 'python')
  end

  -- Find and use virtualenv from pipenv in workspace directory.
  local match = vim.fn.glob(path.join(workspace, 'Pipfile'))
  if match ~= '' then
    local venv = vim.fn.trim(vim.fn.system('PIPENV_PIPFILE=' .. match .. ' pipenv --venv'))
    return path.join(venv, 'bin', 'python')
  end

  -- Fallback to system Python.
  return vim.fn.exepath('python3') or vim.fn.exepath('python') or 'python'
end

require'lspconfig'.pyright.setup {
    on_attach = function()
        require'lsp_signature'.on_attach {
            hint_enable = false,
        }
    end,
    on_init = function(client)
        client.config.settings.python.pythonPath = get_python_path(client.config.root_dir)
    end
}
```

**Comment by cilsat** [comment]
And tweaking @gzbd code for Poetry, substitute the Pipenv-related code in the middle section with:
```lua
  -- Find and use virtualenv via poetry in workspace directory.
  local match = vim.fn.glob(path.join(workspace, 'poetry.lock'))
  if match ~= '' then
    local venv = vim.fn.trim(vim.fn.system('poetry env info -p'))
    return path.join(venv, 'bin', 'python')
  end
```

**Comment by mjlbach** [comment]
Hi all,

I strongly recommend the creation of a python.nvim plugin (akin to nvim-jdtls, nvim-metals, flutter-tools.nvim, etc). LspConfig is meant to be a collection of settings for the clients, but something like managing your python path/active virtual environments is out of scope (would probably benefit from a nice GUI). That being said, if someone tries to write a python.nvim and runs into issues, LMK.

**Comment by HallerPatrick** [comment]
Alright, I was so free and created a repo which now contains the working example of @lithammer : https://github.com/HallerPatrick/py_lsp.nvim.

It serves as a starter to make something usable and nice. Please feel free to contribute :)

**Comment by ajslater** [comment]
Inelegant workaround:

```bash
#!/bin/bash
set -uo pipefail
TOPLEVEL=$(git rev-parse --show-toplevel)
LOCKFILE="$TOPLEVEL/poetry.lock"
if [ -f "$LOCKFILE" ]; then
  poetry run nvim "$@"
else
  nvim "$@"
fi
```

**Comment by bellini666** [comment]
Just my 2 cents in here too, I'm doing this:

```lua
before_init = function(_, config)
    local p
    if vim.env.VIRTUAL_ENV then
        p = lsp_util.path.join(vim.env.VIRTUAL_ENV, "bin", "python3")
    else
        p = utils.find_cmd("python3", ".venv/bin", config.root_dir)
    end
    config.settings.python.pythonPath = p
end,
```

That `func_cmd` comes from my [utils file](https://github.com/bellini666/dotfiles/blob/master/vim/lua/utils.lua#L111) and is a custom implementation of the one suggested by the null-ls developer in [this issue](https://github.com/jose-elias-alvarez/null-ls.nvim/issues/289#issuecomment-953420169).

I use it for flake8/isort/black to find the correct binaries from my poetry installation (which in my case, gets installed in my `.venv` dir): https://github.com/bellini666/dotfiles/blob/master/vim/lua/config/lsp.lua#L184

**Comment by ajslater** [comment]
I've recently switched to null-ls from efm. It's far easier to configure and seems a little faster. Most of these python venv problems can be fixed with null-ls's `{ prefer_local = ".venv/bin" }` configuration. 

But for pyright I implemented @bellini666's solution. While it's a bit of extra code, it works well.

**Comment by Lilja** [comment]
There's one slight issue with the poetry one that's circulating in the thread:

```lua
  -- Find and use virtualenv via poetry in workspace directory.
  local match = vim.fn.glob(path.join(workspace, 'poetry.lock'))
  if match ~= '' then
    local venv = vim.fn.trim(vim.fn.system('poetry env info -p'))
    return path.join(venv, 'bin', 'python')
  end
```

And that it's running the `poetry` command in the CWD of where you are, if you are running multiple nested poetry enviroments like me, then you need to use the `--directory` flag from poetry `1.3.0`.

```lua
  -- Find and use virtualenv via poetry in workspace directory.
  local match = vim.fn.glob(path.join(workspace, 'poetry.lock'))
  if match ~= '' then
    local venv = vim.fn.trim(vim.fn.system('poetry --directory ' .. workspace .. ' env info -p'))
    return path.join(venv, 'bin', 'python')
  end
```

Now, if you have nested poetry environments it will work just fine. Just remember to update `poetry` to `1.3.0` and higher.

**Comment by raffaem** [comment]
So, I'm using pyenv and pylsp

1. The solution by [bellini666](https://github.com/neovim/nvim-lspconfig/issues/500#issuecomment-965824580) is only valid for pyright, as it is based on a configuration found [here](https://github.com/neovim/nvim-lspconfig/blob/master/lua/lspconfig/server_configurations/pyright.lua#L34). The [pylsp config](https://github.com/neovim/nvim-lspconfig/blob/master/lua/lspconfig/server_configurations/pylsp.lua) has no such equivalent.

2. I *think* that pylsp abides the `VIRTUAL_ENV` environmental variables (which is automatically set by pyenv when to point to the correct virtual environment, if you correctly initialize pyenv in your .bashrc as described in the guide). You can test it like that: create two virtual environments, `env1` and `env2`. Install a package, like `tqdm` for example, on `env2` only. Activate `env1`. Open nvim, type `tqdm.`. You will not get autocompletions. Issue `VIRTUAL_ENV=[PATH_TO_ENV2] nvim`, type `tqdm.`. Now you get autocompletion (even if you are technically "inside" `env1`).

But that is only a conjecture by trial and error. Can someone confirm this?

4. The python provider for neovim does not need to correspond to the python in the virtual environment. You can create a virtual environment just for the provider, called for example `neovim`, install `pynvim` on it, and put the following in your `init.lua`: 
```
MYHOME = os.getenv("HOME")
vim.g.python3_host_prog = MYHOME .. "/.pyenv/versions/neovim/bin/python"
```

On a related  note, why everybody is using pyright? Is it not a static type checker only? How you do autocompletions?

**Comment by sblask** [comment]
I spent quite a bit of time trying to find a solution to use a venv specific to an open buffer. Most people seem to do what I used to do, enable the venv before starting nvim, but that doesn't work if you want to have two projects open in the same neovim instance. The fact that `cmd_env` needs to be static appeared to make it unsuitable. The solution is embarrassingly simple - use a relative path for `VIRTUAL_ENV`:

```
local lspconfig = require("lspconfig")
lspconfig["pylsp"].setup({
    cmd_env = {
        VIRTUAL_ENV = ".venv",
    }
})
```

This works because pylsp is started (which isn't super obvious) in `root_dir` (the root of a python project https://github.com/neovim/nvim-lspconfig/blob/master/lua/lspconfig/configs/pylsp.lua#L8). So all you need to do is to create your .venv folder in your python project folder.
